_rules/text-contrast-afw4f7.md

```diff
@@ -0,0 +1,373 @@
+
+```html
+<p style="color: #CCC; background: #fff;" id="p"></p>
+<script>
+	const shadowRoot = document.getElementById('p').attachShadow({ mode: 'open' })
+	shadowRoot.textContent = '<span style="color: #333;">Some text in English</span>'
+</script>
+```
+
+### Failed
+
+#### Failed Example 1
+
+This light grey text has a contrast ratio of 2.3:1 on the white background.
+
+```html
+<p style="color: #AAA; background: white;">
+	Some text in English
+</p>
+```
+
+#### Failed Example 2
+
+This light grey text has a contrast ratio between 1.2:1 and 2.3:1 on the white to blue gradient background.
+
+```html
+<p style="color: #AAA; background: linear-gradient(to right, #FFF, #00F); width: 300px">
+	Some text in English
+</p>
+```
+
+#### Failed Example 3
+
+This light grey text has a contrast ratio between 2.7:1 and 3:1 on the background image.
+
+```html
+<p
+	style="color: #555; height:50px; padding-top:20px; background: black no-repeat -20px -20px url('../test-assets/contrast/black-hole.jpeg');"
+>
+	Black hole sun
+</p>
+```
+
+#### Failed Example 4
+
+This black text with 30% alpha channel has a contrast ratio of 2.1:1 on the white background.
+
+```html
+<p style="color: rgba(0,0,0,.3); background: #FFF">
+	Some text in English
+</p>
+```
+
+#### Failed Example 5
+
+This black text with 30% opacity has a contrast ratio of 2.1:1 on the white background.
+
+```html
+<div style="background: #FFF">
+	<p style="color: #000; opacity: .3">
+		Some text in English
+	</p>
+</div>
+```
+
+#### Failed Example 6
+
+This light grey text has a contrast ratio of 2.3:1 on the white background in a shadow DOM tree.
+
+```html
+<p style="color: #aaa; background: #fff;" id="p"></p>
+<script>
+	const shadowRoot = document.getElementById('p').attachShadow({ mode: 'open' })
+	shadowRoot.textContent = 'Some text in English'
+</script>
+```
+
+#### Failed Example 7
+
+This semi-transparent grey text has a contrast ratio between 2.3:1 and 4.2:1 on the black and white background. The light grey text is compared to the white section of the background and the dark grey text is compared to the black section of the background.
+
+```html
+<style>
+	#backgroundSplit {
+		color: rgba(90, 90, 90, 0.8);
+		background-position: top 0 left 0;
+		background-image: linear-gradient(90deg, transparent, transparent 3.3em, black 3.3em, black 6em);
+		padding: 0 1em;
+	}
+</style>
+<span id="backgroundSplit">
+	Hello world
+</span>
+```
+
+### Inapplicable
+
+#### Inapplicable Example 1
+
+This text is not [visible][] because of `display: none`.
+
+```html
+<p style="display: none">Some invisible text in English</p>
+```
+
+#### Inapplicable Example 2
+
+This text is not [visible][] because it is positioned off screen.
+
+```html
+<p style="position:absolute; top: -999em">Some invisible text in English</p>
+```
+
+#### Inapplicable Example 3
+
+This text is not [visible][] because the foreground color is the same as the background color.
```

Comment:
Is this a WCAG allowable way to "hide" text? The moment a user applies their own styles, or selects content that includes it, the text would be visible.

Reply:
If the assumption is "user styles can impact _any_ CSS" then we cannot ever say _anything_ about CSS. For example, hiding something by means of `display: none` would be just as "invalid" as hiding something by means of color; after all, it won't be hidden if overridden by user styles.

It's a slippery slope. If we need to account for the possibility of user styles, we need to be _very_ specific about which properties we expect users to override.

Reply:
That's a fair comment. However, changing text and background colours is a very common accessibility change, eg. for high contrast or invert modes. Something that could unintentionally reveal such text. 

I was questioning the use of matching text and background colour as an acceptable way to "hide" text, because I don't think it is in keeping with the WCAG guidelines and was commonly used as a method to cheat SEO. There are more acceptable ways to visually hide text while keeping it part of the DOM, and this particular method breaks the minimal contrast guideline. I don't think it is inapplicable at all, and am wondering why it is included.

Reply:
I definitely agree that using identical fore- and background colors as a means of hiding text is questionable. What I was trying to get at is that there's a subtle, but important, difference between this...

> This text should not be considered hidden because user styles can override text color.

...and this:

> This text should not be considered hidden because text color can be assumed to be overridden by user styles.

I'd be perfectly okay with not considering identical fore- and background colors as a means of hiding text for the rules where it makes sense, of course coupled with an assumption that explains _why_.

Reply:
That's a very good point Emma. Thanks for bringing this up. I see where you're coming from, custom styles / colors matter for 1.4.3, as the understanding doc explains:

> source: https://www.w3.org/WAI/WCAG21/Understanding/contrast-minimum.html
> It is a failure if no background color is specified when the text color is specified, because the user's default background color is unknown and cannot be evaluated for sufficient contrast.

I agree with you this is an accessibility problem. But I suspect that in most cases such texts should be treated as "purely decorative", which is an exception under 1.4.3. For the SEO example text for example, it isn't there for an aesthetic purpose, but it doesn't provide information or functionality. Not to the user anyway.

I'll add an assumption to the rule to cover this case. I think there are cases where a 1:1 contrast text fails 1.4.3, but my guess is that usually it doesn't.

Reply:
I would not consider text the same colour as the background to be decorative, unless it was using a drop-shadow or similar to create a visible pattern of some kind. An Inapplicable Example that is clearly demonstrating decorative text would be a different matter.

https://www.w3.org/WAI/WCAG21/Understanding/contrast-minimum.html includes...
"This requirement applies to situations in which images of text were intended to be understood as text. Incidental text, such as in photographs that happen to include a street sign, are not included. Nor is text that for some reason is designed to be invisible to all viewers."

That is the only exception I have found any information about, and I'm not sure this is a good example of that.

Reply:
Perhaps something like:

This text is not fully [visible][] because it is intended to be decorative rather than meaningful, so colour minimal contrast is not necessary.
```html
<div style="color: white; background: white; text-shadow: 2px 2px;" aria-hidden="true">Meaningless white on white text</div>
```

Reply:
Or at least a note that this is not a recommended technique... and an additional example of decorative incidental meaningless text.